### ISSUE TITLE
Supporting web browsers by using a Buffer polyfill

### ISSUE THREAD

**Issue by brickpop** [ethereumjs/merkle-patricia-tree#138]
The standard JS class to represent byte arrays is `Uint8Array`, which NodeJS extends with extra features on the `Buffer` class, used in this repo.

However, the Buffer class is not available on web browsers, so people trying to validate a Merke Proof will not be able to provide valid parameters. 

There is a polyfill available (https://www.npmjs.com/package/buffer) which provides a drop-in replacement that also inherits from `Uint8Array`.

Our particular problem arises when using `BaseTrie.fromProof`, at [which point](https://github.com/ethereumjs/merkle-patricia-tree/blob/master/src/baseTrie.ts#L614-L621) `keccak(nodeValue)` throws an error because the `Uint8Array` value is not a NodeJS `Buffer`.

- Would it be possible to use another `keccak()` implementation that supported `Uint8Arrays`?
- Should `ethereumjs-util` > `keccak()` be updated instead?

The alternative would be for the caller to manually overwrite the global `Buffer` class, which is anything but elegant. Also, depending on a global `Buffer` class that is not explicitly imported, could allow other dependencies to inject a malicious global `Buffer` replacement. 

Thank you!

**Comment by alcuadrado** [comment]
Hey @brickpop, how are you bundling your application?

**Comment by brickpop** [comment]
Hi @alcuadrado 
We typically use Next.js (Webpack) but to avoid encountering the Buffer problem there, our intermediary libraries already use the NPM Buffer version. This allows anyone to import our library and bundle it with their tool of choice. 

So far, the problem arises when running Mocha to test our code. Haven't got any further without the proof validation.

**Comment by holgerd77** [comment]
I just scrolled a bit through the [Buffer usages](https://github.com/ethereumjs/merkle-patricia-tree/search?q=Buffer) in this library, usage patterns are not very complex and the API affected is pretty limited. @alcuadrado would this eventually be a candidate for some first Buffer -> UInt8Array switch? Considering also in particular that we now have the library [in the monorepo](https://github.com/ethereumjs/ethereumjs-vm/tree/master/packages/trie) and can do integrated testing + have some kind of breaking release anyhow by renaming to `@ethereumjs/trie` (and maybe API wise this might not necessarily be (too) breaking, or at least it might be (somewhat) easily possible to remain support for both the existing Buffer and a potential future non-Buffer version). What do you guys think?

//cc @jochem-brouwer @cgewecke

**Comment by cgewecke** [comment]
@brickpop Out of curiosity is your immediate problem getting mocha tests to execute in a browser context? This repo runs passing browser integration tests with karma that [exercise the same section of code, here][1] ...

[1]: https://github.com/ethereumjs/merkle-patricia-tree/blob/01f40d87090a7b6481ee99eb328f69bca5976131/test/proof.spec.ts#L7-L16

**Comment by brickpop** [comment]
@cgewecke not really, the direct development is on a TypeScript library and testing done within the NodeJS context.

We want to support frontend applications beyond Node, so we use a polyfilled `Buffer` instead of the native Node's version. And this will cause `BaseTrie.fromProof` > ethereumjs-util > `keccak()` to throw an error because of the class mismatch. 

Do your browser integration tests inject the Buffer themselves? Otherwise they should not work out of the box:

![image](https://user-images.githubusercontent.com/4456749/104017435-65e0dd80-51b8-11eb-8969-ccb366c9a8b5.png)

**Comment by ryanio** [comment]
> Do your browser integration tests inject the Buffer themselves? Otherwise they should not work out of the box

We use karma-typescript which polyfills with feross/buffer. Prior to that we were using browserify which did the same and was also working.

Are your mocha tests running after the webpack buffer replacement? If MPT and ethereumjs-util are properly bundled together and buffer is replaced it should work.

**Comment by jochem-brouwer** [comment]
@holgerd77 since we want to remove node's `Buffer` since we don't need this rather "extended" version of `Uint8Array` in most (if not all) circumstances, I'd say that this would be a good first candidate to remove the `Buffer` dependency.

**Comment by cgewecke** [comment]
@holgerd77 @jochem-brouwer 

`Buffer` is such a widespread datatype in these libs that I'd like to hear the arguments for making this change again. 

Wrt this issue, a common approach when publishing a lib which works for both Node and the browser is to allow the end user to manage the polyfill (via a bundler) and/or publish a bundled version to CDN to simplify browser availability.

**Comment by alcuadrado** [comment]
I really really really want to remove Node APIs where there's a native counterpart. But for things as widespread as Buffer is really hard to do it. You don't just need to remove it from your code, but also make sure that none of your dependencies use it. If a single one uses it, all of its problems are still present.

Does that mean that we just shouldn't care and accept the situation? I don't think so, but rather than removing it would be an ecosystem-wide effort. Maybe we should do it while switching from BN.js to native big int, which also has some of this characteristics.
